include/fiction/algorithms/physical_design/graph_oriented_layout_design.hpp

```diff
@@ -0,0 +1,1574 @@
+                layout.resize({bb.get_max().x, bb.get_max().y, layout.z()});
+
+                area                     = layout.area();
+                best_solution            = area;
+                improve_current_solution = true;
+
+                fiction::post_layout_optimization_params plo_params{};
+                plo_params.optimize_pos_only = true;
+
+                fiction::post_layout_optimization(layout, plo_params);
+
+                const auto bb_after_plo = fiction::bounding_box_2d(layout);
+                layout.resize({bb_after_plo.get_max().x, bb_after_plo.get_max().y, layout.z()});
+
+                if (layout.area() < best_optimized_solution)
+                {
+                    best_optimized_solution = layout.area();
+
+                    if (ps.verbose)
+                    {
+                        print_placement_info(layout);
+                    }
+
+                    return {{}, layout};
+                }
+
+                return {{}, std::nullopt};
+            }
+
+            if (improve_current_solution && area >= best_solution)
+            {
+                return {{}, std::nullopt};
+            }
+
+            adjust_layout_size(position, layout, ssg, place_info);
+            // check if it's the last position in the current vertex
+            if (idx == (ssg.current_vertex.size() - 1))
+            {
+                if (!valid_layout(layout, ssg, place_info))
+                {
+                    return {{}, std::nullopt};
+                }
+
+                possible_positions = get_possible_positions(layout, ssg, place_info);
+            }
+        }
+
+        return generate_next_positions(possible_positions, layout, ssg);
+    }
+    /**
+     * Initializes the allowed positions for primary inputs (PIs), the cost for each search space graph and the maximum
+     * number of expansions.
+     */
+    void initialize_pis_cost_and_num_expansions() noexcept
+    {
+        static constexpr std::array<pi_locations, 12> pi_locs = {{{true, false},
+                                                                  {false, true},
+                                                                  {true, true},
+                                                                  {true, false},
+                                                                  {false, true},
+                                                                  {true, true},
+                                                                  {true, false},
+                                                                  {false, true},
+                                                                  {true, true},
+                                                                  {true, false},
+                                                                  {false, true},
+                                                                  {true, true}}};
```

Comment:
As mentioned above, this can probably be made a little more meaningful with `enum class` objects